x/gov/keeper.go

```diff
@@ -0,0 +1,415 @@
+package gov
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	wire "github.com/cosmos/cosmos-sdk/wire"
+	"github.com/cosmos/cosmos-sdk/x/bank"
+
+	stake "github.com/cosmos/cosmos-sdk/x/stake"
+)
+
+// Governance Keeper
+type Keeper struct {
+	// The reference to the CoinKeeper to modify balances
+	ck bank.Keeper
+
+	// The reference to the StakeMapper to get information about stakers
+	sk stake.Keeper
+
+	// The (unexposed) keys used to access the stores from the Context.
+	storeKey sdk.StoreKey
+
+	// The wire codec for binary encoding/decoding.
+	cdc *wire.Codec
+
+	// Reserved codespace
+	codespace sdk.CodespaceType
+}
+
+// NewGovernanceMapper returns a mapper that uses go-wire to (binary) encode and decode gov types.
+func NewKeeper(cdc *wire.Codec, key sdk.StoreKey, ck bank.Keeper, sk stake.Keeper, codespace sdk.CodespaceType) Keeper {
+	return Keeper{
+		storeKey:  key,
+		ck:        ck,
+		sk:        sk,
+		cdc:       cdc,
+		codespace: codespace,
+	}
+}
+
+// Returns the go-wire codec.
+func (keeper Keeper) WireCodec() *wire.Codec {
+	return keeper.cdc
+}
+
+// =====================================================
+// Proposals
+
+// Creates a NewProposal
+func (keeper Keeper) NewProposal(ctx sdk.Context, title string, description string, proposalType string) *Proposal {
+	proposal := &Proposal{
+		ProposalID:       keeper.getNewProposalID(ctx),
+		Title:            title,
+		Description:      description,
+		ProposalType:     proposalType,
+		Status:           "DepositPeriod",
+		TotalDeposit:     sdk.Coins{},
+		SubmitBlock:      ctx.BlockHeight(),
+		VotingStartBlock: -1, // TODO: Make Time
+	}
+	keeper.SetProposal(ctx, proposal)
+	keeper.InactiveProposalQueuePush(ctx, proposal)
+	return proposal
+}
+
+// Get Proposal from store by ProposalID
+func (keeper Keeper) GetProposal(ctx sdk.Context, proposalID int64) *Proposal {
+	store := ctx.KVStore(keeper.storeKey)
+	key := []byte(fmt.Sprintf("%d", proposalID) + ":proposal")
+	bz := store.Get(key)
+	if bz == nil {
+		return nil
+	}
+
+	proposal := &Proposal{}
```

Comment:
I implore that we do not use proposal pointers throughout - everything should be a function of the non-pointer variable and then this can be defined in the following way:

```golang
var proposal Proposal 
keeper.cdc.MustUnmarshalBinary(bz, &proposal)
```